runners/apex/src/main/java/org/apache/beam/runners/apex/ApexYarnLauncher.java

```diff
@@ -0,0 +1,383 @@
+ * limitations under the License.
+ */
+package org.apache.beam.runners.apex;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.datatorrent.api.Attribute;
+import com.datatorrent.api.Attribute.AttributeMap;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.StreamingApplication;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Field;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarFile;
+
+import org.apache.apex.api.EmbeddedAppLauncher;
+import org.apache.apex.api.Launcher;
+import org.apache.apex.api.Launcher.AppHandle;
+import org.apache.apex.api.Launcher.LaunchMode;
+import org.apache.apex.api.Launcher.LauncherException;
+import org.apache.apex.api.Launcher.ShutdownMode;
+import org.apache.apex.api.YarnAppLauncher;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.SerializationUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Proxy to launch the YARN application through the hadoop script to run in the
+ * pre-configured environment (class path, configuration, native libraries etc.).
+ *
+ * <p>The proxy takes the DAG and communicates with the Hadoop services to launch
+ * it on the cluster.
+ */
+public class ApexYarnLauncher {
+  private static final Logger LOG = LoggerFactory.getLogger(ApexYarnLauncher.class);
+
+  public AppHandle launchApp(StreamingApplication app) throws IOException {
+
+    List<File> jarsToShip = getYarnDeployDependencies();
+    StringBuilder classpath = new StringBuilder();
+    for (File path : jarsToShip) {
+      if (path.isDirectory()) {
+        File tmpJar = File.createTempFile("beam-runners-apex-", ".jar");
+        createJar(path, tmpJar);
+        tmpJar.deleteOnExit();
+        path = tmpJar;
+      }
+      if (classpath.length() != 0) {
+        classpath.append(':');
+      }
+      classpath.append(path.getAbsolutePath());
+    }
+
+    EmbeddedAppLauncher<?> embeddedLauncher = Launcher.getLauncher(LaunchMode.EMBEDDED);
+    DAG dag = embeddedLauncher.getDAG();
```

Comment:
Is this functionally equivalent to `new LogicalPlan()`? Why is it necessary to go through creating embedded Launcher instance instead of exposing createDAG() method on the Launcher class?